add JSON_NUMERIC_CHECK to json encode options

devorama opened the pull request:
linked to 
https://github.com/codeigniter4/CodeIgniter4/pull/3286

This allows encoding to make numeric data more true to its type (not enclose them with "", making the receiving system identify
the types easier.

Note making the options variable configurable might be handy in the future depending on the needs of the users
people they interface with they might need additional options added here

**Checklist:**
- [ ] Securely signed commits
- [ ] Component(s) with PHPdocs
- [ ] Unit testing, with >80% coverage
- [ ] User guide updated
- [ ] Conforms to style guide

michalsn commented:
Adding a numeric check just like that may potentially break someone's code. Developers may already use code like: 
```php
if ($foo->bar === '1')
``` 

But I like your idea of making this configurable. Would you like to work on this PR and add something like `$formatterOptions` array to the `Config/Format.php` file? We would use it like:
```php
$config = new Filter();
$options = $config->formatterOptions['application/json'] ?? JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES | JSON_PARTIAL_OUTPUT_ON_ERROR
```
I'm not really sure about naming the keys for this new variable so if anybody has a different idea, I'm all ears.

devorama commented:
Hi there @michalsn  , I will do this asap. Will look at the current names being used and see if I can add something that appropiate.

> But I like your idea of making this configurable. Would you like to work on this PR and add something like `$formatterOptions` array to the `Config/Format.php` file? We would use it like:

devorama commented:
@michalsn  I added the config value and also the code for XML and JSON Formatter, for XML I added the additional option to be added as per the php docs or default to the phpdocs default.

One thing to note is I did not make it to overwrite the default options but to add to it, the reason for this is if you look at
JSONFormatter the JSON_PARTIAL_OUTPUT_ON_ERROR is actually important for error checking afterwards in the same function, thus allowing a total override of default options might break actual logic for that reason I believed it saver to make it a add to current option only. The default for the settings are all null.

michalsn commented:
The whole sense of making it configurable is that we should have the ability to change default flags. So please add all current flags to the config file as default.

You can check later in the code if `JSON_PARTIAL_OUTPUT_ON_ERROR` is set via bitwise operator and add it if needed.
```php
if (! ($options & JSON_PARTIAL_OUTPUT_ON_ERROR))
// add it to options
```

michalsn commented:
Actually I don't think that we have to check it at all (options) - we can just add it and it should work fine (even if it's already present).

johpromed commented:
Hi there. The logic I have is as follow, basically it will do what you suggest in your last post. But based on wanting to give control to the user to select the encoding options (basically if they want JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES )
I will move all but JSON_PARTIAL_OUTPUT_ON_ERROR  to be in the Config/Format.php , then  the result will be achieved what you are aiming for.

```
		$options = $config->formatterOptions['application/json'] ?? 0;
		$options = $options | JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES | JSON_PARTIAL_OUTPUT_ON_ERROR ;

```
> Actually I don't think that we have to check it at all (options) - we can just add it and it should work fine (even if it's already present).

michalsn commented:
Just update your code. If I will have any requests for changes I will let you know :)

devorama commented:
Oops used my company account there :), will do the chagne now, just finished testing it and it seems to work :)

> Just update your code. If I will have any requests for changes I will let you know :)

devorama commented:
@michalsn  K changes made, should work now.

michalsn commented on app/Config/Format.php:
```diff
@@ -39,7 +39,23 @@ class Format extends BaseConfig
+	|
+	*/
+	public $formatterOptions  = [
+		'application/json' => JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES ,
```
```suggestion
		'application/json' => JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES,
```

michalsn commented on system/Format/JSONFormatter.php:
```diff
@@ -56,7 +56,8 @@ class JSONFormatter implements FormatterInterface
 	public function format($data)
 	{
-		$options = JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES | JSON_PARTIAL_OUTPUT_ON_ERROR;
+		$options = $config->formatterOptions['application/json'] ?? 0;
```
```suggestion
		$options = $config->formatterOptions['application/json'] ?? JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES;
```
This way we can introduce this change with new configuration variable really easy.

michalsn commented on app/Config/Format.php:
```diff
@@ -39,7 +39,23 @@ class Format extends BaseConfig
+	*/
+	public $formatterOptions  = [
+		'application/json' => JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES ,
+		'application/xml'  => null,
+		'text/xml'         => null,
```
Please make it default to `0`.

michalsn commented on system/Format/JSONFormatter.php:
```diff
@@ -56,7 +56,8 @@ class JSONFormatter implements FormatterInterface
 	{
-		$options = JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES | JSON_PARTIAL_OUTPUT_ON_ERROR;
+		$options = $config->formatterOptions['application/json'] ?? 0;
+		$options = $options | JSON_PARTIAL_OUTPUT_ON_ERROR ;
```
```suggestion
		$options = $options | JSON_PARTIAL_OUTPUT_ON_ERROR;
```

devorama commented:
@michalsn changes made,will try and enforce the coding standard more with next changes.

michalsn commented on system/Format/XMLFormatter.php:
```diff
@@ -66,7 +66,8 @@ public function format($data)
 
-		$output = new \SimpleXMLElement('<?xml version="1.0"?><response></response>');
+		$options = $config->formatterOptions['application/xml'] ?? 0;
+		$output = new \SimpleXMLElement('<?xml version="1.0"?><response></response>',$options);
```
```suggestion
		$output = new \SimpleXMLElement('<?xml version="1.0"?><response></response>', $options);
```
Sorry, I missed this one.

michalsn commented:
@devorama No worries. Just one last change and I think we're good.

devorama commented:
@michalsn  all done :)

paulbalandan commented on app/Config/Format.php:
```diff
@@ -39,7 +39,23 @@ class Format extends BaseConfig
+	| For each mime type, list the additional options that should be used. 
+	| Set to null by default
+	| Note this will  add to the default options not replace it.
+	|
```
Please change the comment. The $options are no longer additions but rather the defaults already.

devorama commented:
@michalsn  and @paulbalandan  Made the changes as requested , sorry for the config slip up.
